File: solutions/2461. Maximum Sum of Distinct Subarrays With Length K.cpp
class Solution {
public:
    long long maximumSubarraySum(vector<int>& nums, int k) {
        //initialization
        unordered_map<int, int> seen;
        unordered_set<int> duplicate;
        long long val=0;
        for (int i=0; i<k;i++){
            val += nums[i];
            if (seen.contains(nums[i])) {
                seen[nums[i]] += 1;
                duplicate.insert(nums[i]);
            } else {
                seen[nums[i]] = 1;
            }
        }
        long long ans = duplicate.size() == 0 ? val : 0;
        
        for (int i = k; i < nums.size();i++) {
            seen[nums[i-k]] -= 1;
            if (duplicate.contains(nums[i-k]) && seen[nums[i-k]] <= 1) {
                duplicate.erase(nums[i-k]);
            }
            if (seen.contains(nums[i])) {
                seen[nums[i]] += 1;
            } else {
                seen[nums[i]] = 1;
            }
            
            if (seen[nums[i]] > 1) {
                duplicate.insert(nums[i]);
            }
            
            val -= nums[i-k];
            val += nums[i];
            
            if (duplicate.size() == 0) ans = max(ans, val);
        }
        return ans;
    }
};

File: solutions/2461. Maximum Sum of Distinct Subarrays With Length K.py
from collections import Counter
class Solution:
    def maximumSubarraySum(self, nums: List[int], k: int) -> int:
        seen = Counter(nums[:k]) #unordered_map
        val = sum(nums[:k])
        duplicate = set() #unordered_set
        if len(seen) < k:
            ans = 0
            duplicate.update([s for s in seen if seen[s] > 1])
        else:
            ans = val
        
        # [1,1,1,2,3]
        for i in range(k, len(nums)):
            seen[nums[i-k]] -= 1
            if nums[i-k] in duplicate and seen[nums[i-k]] <= 1:
                duplicate.remove(nums[i-k])
                
            seen[nums[i]] += 1
            if seen[nums[i]] > 1:
                duplicate.add(nums[i])
            
            val -= nums[i-k]
            val += nums[i]
            
            if len(duplicate) == 0: ans = max(ans,val)
        
        return ans
